### PR TITLE
Eka 132: Encryption of data 

### PR DESCRIPTION
**Pull request by Nexengineer**

### CODE REVIEW

File: test/In.ProjectEKA.HipServiceTest/DataFlow/Builder/KeyMaterialBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+namespace In.ProjectEKA.HipServiceTest.DataFlow.Builder
+{
+    using HipService.DataFlow;
+
+    public class KeyMaterialBuilder
+    {
+        private string CryptoAlg;
+        private string Curve;
+        private KeyStructure DhPublicKey;
+        private string RandomKey;
```
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/DataFlow/Builder/KeyMaterialBuilder.cs]: Codacy Issue found: [Remove unassigned field 'RandomKey', or set its value.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5073772)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 8
           

Complexity increasing per file
==============================
- src/In.ProjectEKA.HipService/DataFlow/CryptoHelper/AesGcmEncryptor.cs  1
- test/In.ProjectEKA.HipServiceTest/DataFlow/Builder/KeyMaterialBuilder.cs  1
- test/In.ProjectEKA.HipServiceTest/DataFlow/CryptoHelperTest.cs  1
- src/In.ProjectEKA.HipLibrary/Patient/Model/KeyStructure.cs  1
- src/In.ProjectEKA.HipService/DataFlow/KeyMaterial.cs  1
- src/In.ProjectEKA.HipService/DataFlow/CryptoHelper/CryptoHelper.cs  2
- src/In.ProjectEKA.HipService/DataFlow/KeyStructure.cs  1
- src/In.ProjectEKA.HipLibrary/Patient/Model/KeyMaterial.cs  1
         

Clones added
============
- src/In.ProjectEKA.HipService/DataFlow/KeyMaterial.cs  1
- src/In.ProjectEKA.HipLibrary/Patient/Model/KeyMaterial.cs  1
         

Clones removed
==============
+ test/In.ProjectEKA.HipServiceTest/DataFlow/Builder/TestBuilder.cs  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5073772&bid=16692126)

File: test/In.ProjectEKA.HipServiceTest/DataFlow/Builder/KeyMaterialBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+namespace In.ProjectEKA.HipServiceTest.DataFlow.Builder
+{
+    using HipService.DataFlow;
+
+    public class KeyMaterialBuilder
+    {
+        private string CryptoAlg;
+        private string Curve;
+        private KeyStructure DhPublicKey;
```
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/DataFlow/Builder/KeyMaterialBuilder.cs]: Codacy Issue found: [Remove unassigned field 'DhPublicKey', or set its value.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5073772)

File: test/In.ProjectEKA.HipServiceTest/DataFlow/Builder/KeyMaterialBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
+namespace In.ProjectEKA.HipServiceTest.DataFlow.Builder
+{
+    using HipService.DataFlow;
+
+    public class KeyMaterialBuilder
+    {
+        private string CryptoAlg;
```
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/DataFlow/Builder/KeyMaterialBuilder.cs]: Codacy Issue found: [Remove unassigned field 'CryptoAlg', or set its value.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5073772)

File: test/In.ProjectEKA.HipServiceTest/DataFlow/Builder/DataRequestBuilder.cs [hunk 4]
```diff
@@ -10,10 +10,11 @@ internal class DataRequestBuilder
         private HiDataRange DataRange;
         private IEnumerable<HiType> HiType;
         private string TransactionId;
+        private KeyMaterial KeyMaterial;
```
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/DataFlow/Builder/DataRequestBuilder.cs]: Codacy Issue found: [Remove unassigned field 'KeyMaterial', or set its value.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5073772)

File: src/In.ProjectEKA.HipService/DataFlow/CryptoHelper/CryptoHelper.cs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+            rngCryptoServiceProvider.GetBytes(randomBytes);
+            return GetBase64FromByte(randomBytes);   
+        }
+
+        private static IEnumerable<byte> XorOfRandom(string randomKeySender, string randomKeyReceiver)
+        {
+            var randomKeySenderBytes = GetByteFromBase64(randomKeySender).ToArray();
+            var randomKeyReceiverBytes = GetByteFromBase64(randomKeyReceiver).ToArray();
+            var sb = new byte[randomKeyReceiverBytes.Length];
+            for (var i = 0; i < randomKeySenderBytes.Length; i++)
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/DataFlow/CryptoHelper/CryptoHelper.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'for' block.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5073772)

File: src/In.ProjectEKA.HipService/DataFlow/DataFlowClient.cs [hunk 6]
```diff
@@ -5,41 +5,61 @@ namespace In.ProjectEKA.HipService.DataFlow
...
+                             EncryptData(dataRequest.KeyMaterial, serializer.SerializeToString(bundle), 
+                                 senderKeyPair, randomKeySender),
                             "application/json",
                             "MD5"))
                         .ToList();
-                    await SendDataToHiu(new DataResponse(dataRequest.TransactionId, healthRecordEntries),
+                    var keyMaterial = new KeyMaterial(dataRequest.KeyMaterial.CryptoAlg, dataRequest.KeyMaterial.Curve,
+                        new KeyStructure("", "", cryptoHelper.GetPublicKey(senderKeyPair)),
+                        randomKeySender);                    
+                    await SendDataToHiu(new DataResponse(dataRequest.TransactionId, healthRecordEntries, keyMaterial),
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/DataFlow/DataFlowClient.cs]: Codacy Issue found: [Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5073772)

File: test/In.ProjectEKA.HipServiceTest/DataFlow/Builder/KeyMaterialBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+        private string Curve;
```
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/DataFlow/Builder/KeyMaterialBuilder.cs]: Codacy Issue found: [Remove unassigned field 'Curve', or set its value.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5073772)

File: test/In.ProjectEKA.HipServiceTest/DataFlow/Builder/TestBuilder.cs [hunk 8]
```diff
@@ -44,5 +55,24 @@ internal static Consent Consent()
                 ConsentStatus.GRANTED
             );
         }
+        
+        internal static DataRequestLib DataFlowRequest()
+        {
+            var faker = new Faker();
```
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/DataFlow/Builder/TestBuilder.cs]: Codacy Issue found: [Rename 'faker' which hides the field with the same name.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5073772)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- src/In.ProjectEKA.HipLibrary/Patient/Model/KeyStructure.cs  1
- test/In.ProjectEKA.HipServiceTest/DataFlow/CryptoHelperTest.cs  1
- src/In.ProjectEKA.HipService/DataFlow/CryptoHelper/AesGcmEncryptor.cs  1
- src/In.ProjectEKA.HipService/DataFlow/CryptoHelper/CryptoHelper.cs  2
- src/In.ProjectEKA.HipService/DataFlow/KeyMaterial.cs  1
- src/In.ProjectEKA.HipLibrary/Patient/Model/KeyMaterial.cs  1
- src/In.ProjectEKA.HipService/DataFlow/KeyStructure.cs  1
         

Clones added
============
- src/In.ProjectEKA.HipService/DataFlow/KeyMaterial.cs  1
- src/In.ProjectEKA.HipLibrary/Patient/Model/KeyMaterial.cs  1
         

Clones removed
==============
+ test/In.ProjectEKA.HipServiceTest/DataFlow/Builder/TestBuilder.cs  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5073772&bid=16692126)

File: test/In.ProjectEKA.HipServiceTest/DataFlow/Builder/DataRequestBuilder.cs (hunk 4 above)
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/DataFlow/Builder/DataRequestBuilder.cs]: Codacy Issue found: [Remove unassigned field 'KeyMaterial', or set its value.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5073772)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 2
           

Complexity increasing per file
==============================
- src/In.ProjectEKA.HipService/DataFlow/KeyStructure.cs  1
- test/In.ProjectEKA.HipServiceTest/DataFlow/CryptoHelperTest.cs  1
- src/In.ProjectEKA.HipService/DataFlow/CryptoHelper/AesGcmEncryptor.cs  1
- src/In.ProjectEKA.HipLibrary/Patient/Model/KeyStructure.cs  1
- src/In.ProjectEKA.HipLibrary/Patient/Model/KeyMaterial.cs  1
- src/In.ProjectEKA.HipService/DataFlow/CryptoHelper/CryptoHelper.cs  2
- src/In.ProjectEKA.HipService/DataFlow/KeyMaterial.cs  1
         

Clones added
============
- src/In.ProjectEKA.HipLibrary/Patient/Model/KeyMaterial.cs  1
- src/In.ProjectEKA.HipService/DataFlow/KeyMaterial.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5073772&bid=16692126)

File: test/In.ProjectEKA.HipServiceTest/DataFlow/Builder/DataRequestBuilder.cs (hunk 4 above)
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/DataFlow/Builder/DataRequestBuilder.cs]: Codacy Issue found: [Remove unassigned field 'KeyMaterial', or set its value.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5073772)

File: test/In.ProjectEKA.HipServiceTest/DataFlow/DataFlowClientTest.cs [hunk 9]
```diff
@@ -6,21 +6,31 @@ namespace In.ProjectEKA.HipServiceTest.DataFlow
...
-    using Moq;
-    using Moq.Protected;
     using Optional;
     using Xunit;
     using Task = System.Threading.Tasks.Task;
 
     [Collection("Queue Listener Tests")]
     public class DataFlowClientTest
     {
+        private readonly Collect collect = new Collect("observation.json");
```
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/DataFlow/DataFlowClientTest.cs]: Codacy Issue found: [Remove the unused private field 'collect'.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5073772)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- src/In.ProjectEKA.HipLibrary/Patient/Model/KeyMaterial.cs  1
- src/In.ProjectEKA.HipLibrary/Patient/Model/KeyStructure.cs  1
- src/In.ProjectEKA.HipService/DataFlow/CryptoHelper/AesGcmEncryptor.cs  1
- src/In.ProjectEKA.HipService/DataFlow/KeyStructure.cs  1
- src/In.ProjectEKA.HipService/DataFlow/KeyMaterial.cs  1
- test/In.ProjectEKA.HipServiceTest/DataFlow/CryptoHelperTest.cs  1
- src/In.ProjectEKA.HipService/DataFlow/CryptoHelper/CryptoHelper.cs  2
         

Clones added
============
- src/In.ProjectEKA.HipLibrary/Patient/Model/KeyMaterial.cs  1
- src/In.ProjectEKA.HipService/DataFlow/KeyMaterial.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5073772&bid=16692126)

File: src/In.ProjectEKA.HipLibrary/Patient/Model/KeyStructure.cs [hunk 10]
```diff
@@ -0,0 +1,16 @@
+namespace In.ProjectEKA.HipLibrary.Patient.Model
+{
+    public class KeyStructure
+    {
+        public string Expiry { get; }
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipLibrary/Patient/Model/KeyStructure.cs]: If it's datetime, can you use DateTime only.

File: src/In.ProjectEKA.HipService/DataFlow/CryptoHelper/AesGcmEncryptor.cs [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+                var retLen = cipher.ProcessBytes
+                               (plainBytes, 0, plainBytes.Length, encryptedBytes, 0);
+                cipher.DoFinal(encryptedBytes, retLen);
+                encryptedString = Convert.ToBase64String(encryptedBytes, Base64FormattingOptions.None);
+            }
+            catch (Exception ex)
+            {
+                Log.Error(ex.Message);
+                Log.Fatal(ex);
+            }
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/DataFlow/CryptoHelper/AesGcmEncryptor.cs]: Are we thinking to send empty string if we failed to encrypt the data?

File: src/In.ProjectEKA.HipService/DataFlow/CryptoHelper/CryptoHelper.cs [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+    using Newtonsoft.Json;
+    using Org.BouncyCastle.Crypto;
+    using Org.BouncyCastle.Crypto.Digests;
+    using Org.BouncyCastle.Crypto.EC;
+    using Org.BouncyCastle.Crypto.Generators;
+    using Org.BouncyCastle.Crypto.Parameters;
+    using Org.BouncyCastle.Security;
+    using Encoder = Org.BouncyCastle.Utilities.Encoders;
+
+    public class CryptoHelper: ICryptoHelper
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/DataFlow/CryptoHelper/CryptoHelper.cs]: Why we need an interface? And don't call it as Helper.

File: src/In.ProjectEKA.HipService/DataFlow/CryptoHelper/CryptoHelper.cs [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+            var aesKey = new byte[32];
+            hkdfBytesGenerator.GenerateBytes(aesKey, 0, 32);
+            try
+            {
+                return AesGcmEncryptor.EncryptDataUseAesGcm(json, aesKey, iv.ToArray());
+            }
+            catch (Exception e)
+            {
+                Log.Error("Error Occured while encryption {exception}: ",e);
+                return "";
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/DataFlow/CryptoHelper/CryptoHelper.cs]: If there are failure when encrypting, then you should use Optional<String>

File: src/In.ProjectEKA.HipService/DataFlow/CryptoHelper/CryptoHelper.cs (hunk 12 above)
Comment (Nexengineer) [reply src/In.ProjectEKA.HipService/DataFlow/CryptoHelper/CryptoHelper.cs]: Will change the name. Interface is required for testing dataflow client

File: src/In.ProjectEKA.HipService/DataFlow/CryptoHelper/CryptoHelper.cs [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+            return Encoder.Base64.Decode(value);
+        }
+
+        public string GenerateRandomKey()
+        {
+            var rngCryptoServiceProvider = new RNGCryptoServiceProvider();
+            var randomBytes = new byte[32];
+            rngCryptoServiceProvider.GetBytes(randomBytes);
+            return GetBase64FromByte(randomBytes);   
+        }
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/DataFlow/CryptoHelper/CryptoHelper.cs]: Keys generation can be part of different class.

File: src/In.ProjectEKA.HipService/DataFlow/CryptoHelper/ICryptoHelper.cs [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+    using Org.BouncyCastle.Crypto;
+
+    public interface ICryptoHelper
+    {
+        AsymmetricCipherKeyPair GenerateKeyPair(string curveName, string algorithm);
+        string GetPublicKey(AsymmetricCipherKeyPair senderKeyPair);
+        string EncryptData(string receiverPublicKey, AsymmetricCipherKeyPair senderKeyPair, string content,
+            string randomKeySender, string randomKeyReceiver, string curveName, string algorithm);
+        string GenerateRandomKey();
+    }
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/DataFlow/CryptoHelper/ICryptoHelper.cs]: Not required I guess.

File: src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs [hunk 16]
```diff
@@ -34,7 +34,7 @@ public class DataFlow : IDataFlow
             }
 
             var dataRequest = new DataRequest(consent.ConsentArtefact.CareContexts, request.HiDataRange,
-                request.CallBackUrl, consent.ConsentArtefact.HiTypes, request.TransactionId);
+                request.CallBackUrl, consent.ConsentArtefact.HiTypes, request.TransactionId, request.KeyMaterial);
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs]: format this properly.

File: src/In.ProjectEKA.HipService/DataFlow/DataFlowClient.cs [hunk 17]
```diff
@@ -6,41 +6,61 @@ namespace In.ProjectEKA.HipService.DataFlow
...
+    using Org.BouncyCastle.Crypto;
 
     public class DataFlowClient
     {
         private readonly ICollect collect;
         private readonly HttpClient httpClient;
+        private readonly ICryptoHelper cryptoHelper;
 
-        public DataFlowClient(ICollect collect, HttpClient httpClient)
+        public DataFlowClient(ICollect collect, HttpClient httpClient, ICryptoHelper cryptoHelper)
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/DataFlow/DataFlowClient.cs]:
```suggestion
        public DataFlowClient(ICollect collect, HttpClient httpClient, Encryptor cryptoHelper)
```

File: src/In.ProjectEKA.HipService/DataFlow/DataFlowClient.cs [hunk 18]
```diff
@@ -6,41 +6,61 @@ namespace In.ProjectEKA.HipService.DataFlow
...
-                    await SendDataToHiu(new DataResponse(dataRequest.TransactionId, healthRecordEntries),
-                        dataRequest.CallBackUrl);
+                    var keyMaterial = new KeyMaterial(dataRequest.KeyMaterial.CryptoAlg, dataRequest.KeyMaterial.Curve,
+                        new KeyStructure("", "", cryptoHelper.GetPublicKey(senderKeyPair)),
+                        randomKeySender);                    
+                    await SendDataToHiu(new DataResponse(dataRequest.TransactionId, healthRecordEntries, keyMaterial),
+                        dataRequest.CallBackUrl).ConfigureAwait(false);
                     return Task.CompletedTask;
                 });
         }
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/DataFlow/DataFlowClient.cs]: This needs to be refactored, can you come here? :)

File: src/In.ProjectEKA.HipLibrary/Patient/Model/KeyStructure.cs (hunk 10 above)
Comment (Nexengineer) [reply src/In.ProjectEKA.HipLibrary/Patient/Model/KeyStructure.cs]: It is defined as date time in contract

File: src/In.ProjectEKA.HipService/DataFlow/CryptoHelper/ICryptoHelper.cs (hunk 15 above)
Comment (Nexengineer) [reply src/In.ProjectEKA.HipService/DataFlow/CryptoHelper/ICryptoHelper.cs]: required for testing

File: src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs (hunk 16 above)
Comment (Nexengineer) [reply src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs]: Done

File: src/In.ProjectEKA.HipService/DataFlow/DataFlowClient.cs (hunk 17 above)
Comment (Nexengineer) [reply src/In.ProjectEKA.HipService/DataFlow/DataFlowClient.cs]: Done

File: test/In.ProjectEKA.HipServiceTest/DataFlow/DataEntryFactoryTest.cs [hunk 19]
```diff
@@ -3,19 +3,26 @@ namespace In.ProjectEKA.HipServiceTest.DataFlow
...
     using Microsoft.Extensions.Options;
     using Moq;
     using Optional;
+    using Org.BouncyCastle.Crypto;
     using Xunit;
 
     public class DataEntryFactoryTest
     {
+        private static Mock<IEncryptor> encryptor = new Mock<IEncryptor>();
+        private readonly string curve = "curve25519";
```
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/DataFlow/DataEntryFactoryTest.cs]: Codacy Issue found: [Remove the unused private field 'curve'.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5073772)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           

Complexity increasing per file
==============================
- src/In.ProjectEKA.HipService/DataFlow/EncryptedEntries.cs  1
- src/In.ProjectEKA.HipLibrary/Patient/Model/KeyStructure.cs  1
- test/In.ProjectEKA.HipServiceTest/DataFlow/EncryptorHelperTest.cs  1
- src/In.ProjectEKA.HipService/DataFlow/Encryptor/Encryptor.cs  2
- src/In.ProjectEKA.HipService/DataFlow/KeyMaterial.cs  1
- src/In.ProjectEKA.HipService/DataFlow/Encryptor/EncryptorHelper.cs  2
- src/In.ProjectEKA.HipService/DataFlow/DataEntryFactory.cs  2
- src/In.ProjectEKA.HipLibrary/Patient/Model/KeyMaterial.cs  1
- src/In.ProjectEKA.HipService/DataFlow/KeyStructure.cs  1
- test/In.ProjectEKA.HipServiceTest/DataFlow/EncryptorTest.cs  1
         

Clones added
============
- src/In.ProjectEKA.HipService/DataFlow/KeyMaterial.cs  1
- src/In.ProjectEKA.HipLibrary/Patient/Model/KeyMaterial.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5073772&bid=16692126)

File: test/In.ProjectEKA.HipServiceTest/DataFlow/DataFlowMessageHandlerTest.cs [hunk 20]
```diff
@@ -13,6 +13,8 @@ namespace In.ProjectEKA.HipServiceTest.DataFlow
 
     public class DataFlowMessageHandlerTest
     {
+        private readonly string curve = "curve25519";
+        private readonly string algorithm = "ECDH";
```
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/DataFlow/DataFlowMessageHandlerTest.cs]: Codacy Issue found: [Remove the unused private field 'algorithm'.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5073772)

File: test/In.ProjectEKA.HipServiceTest/DataFlow/DataFlowMessageHandlerTest.cs [hunk 21]
```diff
@@ -13,6 +13,8 @@ namespace In.ProjectEKA.HipServiceTest.DataFlow
 
     public class DataFlowMessageHandlerTest
     {
+        private readonly string curve = "curve25519";
```
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/DataFlow/DataFlowMessageHandlerTest.cs]: Codacy Issue found: [Remove the unused private field 'curve'.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5073772)

File: test/In.ProjectEKA.HipServiceTest/DataFlow/EncryptorTest.cs [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+            var result = encryptor.EncryptData(keyMaterial, senderKeyPair, new Faker().Random.Word(), randomKeySecond);
+
+            result.HasValue.Should().BeTrue();
+            result.MatchSome(data => data.Length.Should().BePositive());
+        }
+        
+        [Fact]
+        private void ShouldNotEncryptData()
+        {
+            var faker = new Faker();
```
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/DataFlow/EncryptorTest.cs]: Codacy Issue found: [Remove the unused local variable 'faker'.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5073772)

File: test/In.ProjectEKA.HipServiceTest/DataFlow/DataEntryFactoryTest.cs [hunk 23]
```diff
@@ -3,19 +3,26 @@ namespace In.ProjectEKA.HipServiceTest.DataFlow
...
+        private readonly string algorithm = "ECDH";
```
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/DataFlow/DataEntryFactoryTest.cs]: Codacy Issue found: [Remove the unused private field 'algorithm'.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5073772)

File: test/In.ProjectEKA.HipServiceTest/DataFlow/DataEntryFactoryTest.cs (hunk 19 above)
Comment (Nexengineer) [reply test/In.ProjectEKA.HipServiceTest/DataFlow/DataEntryFactoryTest.cs]: done

File: test/In.ProjectEKA.HipServiceTest/DataFlow/DataFlowMessageHandlerTest.cs (hunk 20 above)
Comment (Nexengineer) [reply test/In.ProjectEKA.HipServiceTest/DataFlow/DataFlowMessageHandlerTest.cs]: done

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/In.ProjectEKA.HipLibrary/Patient/Model/KeyMaterial.cs  1
- src/In.ProjectEKA.HipService/DataFlow/KeyMaterial.cs  1
- src/In.ProjectEKA.HipService/DataFlow/Encryptor/EncryptorHelper.cs  2
- src/In.ProjectEKA.HipService/DataFlow/Encryptor/Encryptor.cs  2
- test/In.ProjectEKA.HipServiceTest/DataFlow/EncryptorHelperTest.cs  1
- src/In.ProjectEKA.HipService/DataFlow/KeyStructure.cs  1
- src/In.ProjectEKA.HipService/DataFlow/EncryptedEntries.cs  1
- src/In.ProjectEKA.HipService/DataFlow/DataEntryFactory.cs  2
- test/In.ProjectEKA.HipServiceTest/DataFlow/EncryptorTest.cs  1
- src/In.ProjectEKA.HipLibrary/Patient/Model/KeyStructure.cs  1
         

Clones added
============
- src/In.ProjectEKA.HipLibrary/Patient/Model/KeyMaterial.cs  1
- src/In.ProjectEKA.HipService/DataFlow/KeyMaterial.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5073772&bid=16692126)

File: src/In.ProjectEKA.HipLibrary/Patient/Model/KeyStructure.cs (hunk 10 above)
Comment (ganesanarun) [reply src/In.ProjectEKA.HipLibrary/Patient/Model/KeyStructure.cs]: Json does not support DateTime :)

File: src/In.ProjectEKA.HipService/DataFlow/DataFlowMessageHandler.cs [hunk 24]
```diff
@@ -21,9 +23,15 @@ public class DataFlowMessageHandler
...
+            var sentKeyMaterial = dataRequest.KeyMaterial;
+
             var data = await collect.CollectData(dataRequest);
-            var entries = dataEntryFactory.Process(data);
-            dataFlowClient.SendDataToHiu(dataRequest, entries);
+            var entries = dataEntryFactory.Process(data, sentKeyMaterial);
+            entries.MatchSome(encryptedEntries => 
+                    dataFlowClient.SendDataToHiu(dataRequest,
+                        encryptedEntries.Entries,
+                        encryptedEntries.KeyMaterial))
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/DataFlow/DataFlowMessageHandler.cs]: format this properly.

**Comment by ganesanarun** [comment]
Format everything properly before you merge it. :)

File: src/In.ProjectEKA.HipService/DataFlow/DataFlowMessageHandler.cs (hunk 24 above)
Comment (Nexengineer) [reply src/In.ProjectEKA.HipService/DataFlow/DataFlowMessageHandler.cs]: Done

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- test/In.ProjectEKA.HipServiceTest/DataFlow/EncryptorHelperTest.cs  1
- src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs  1
- src/In.ProjectEKA.HipService/DataFlow/EncryptedEntries.cs  1
- src/In.ProjectEKA.HipService/DataFlow/DataEntryFactory.cs  2
- src/In.ProjectEKA.HipService/DataFlow/Encryptor/Encryptor.cs  2
- src/In.ProjectEKA.HipService/DataFlow/KeyStructure.cs  1
- src/In.ProjectEKA.HipLibrary/Patient/Model/KeyStructure.cs  1
- src/In.ProjectEKA.HipService/DataFlow/Encryptor/EncryptorHelper.cs  2
- test/In.ProjectEKA.HipServiceTest/DataFlow/EncryptorTest.cs  1
- src/In.ProjectEKA.HipLibrary/Patient/Model/KeyMaterial.cs  1
- src/In.ProjectEKA.HipService/DataFlow/KeyMaterial.cs  1
         

Clones added
============
- src/In.ProjectEKA.HipLibrary/Patient/Model/KeyMaterial.cs  1
- src/In.ProjectEKA.HipService/DataFlow/KeyMaterial.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5073772&bid=16692126)

**Pull request merged**